### PR TITLE
[2.7.x] Use error.invalid when form binding fails and messages are empty

### LINKED ISSUE #8318: errorsAsJson with mocked MessagesApi throws java.util.NoSuchElementException (TestCase provided)
### Play Version (2.5.x / etc)
2.6.11

### API (Scala / Java / Neither / Both)
Java

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)
Windows 7

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)
JDK 9.0.4

### Library Dependencies
N/A

### Expected Behavior
No exception is thrown

### Actual Behavior
I want my tests to be language independent, thus I want `form.errorsAsJson()` to return error keys instead of translations.
In order to do that I mocked MessagesApi as documentation specifies:
```
public class MyTest extends WithApplication {

  private MessagesApi emptyMessagesApi() {
    return new play.i18n.MessagesApi(new play.api.i18n.DefaultMessagesApi(
        Collections.singletonMap(Lang.defaultLang().code(), Collections.emptyMap()),
        new play.api.i18n.DefaultLangs().asJava()));
  }

  @Override
  protected Application provideApplication() {
    GuiceApplicationBuilder builder = new GuiceApplicationBuilder()
        .in(Environment.simple())
        .overrides(bindings.add(Bindings.bind(MessagesApi.class).toInstance(emptyMessagesApi())));

    Guice.createInjector(builder.applicationModule()).injectMembers(this);
    Application application = builder.build();
    return application;
  }

  public Application getApplication() {
    return app;
  }

//here test making call to controller
}
```

While calls to `messages.at("someKey")` work fine, trying to call `form.errorsAsJson()` throws:
```
[error] Test controllers.MyControllerTest.shouldValidateRequiredFields failed: java.util.NoSuchElementException: null, took 0.881 sec
[error]     at java.util.ArrayList$Itr.next(ArrayList.java:896)
[error]     at scala.collection.convert.Wrappers$JIteratorWrapper.next(Wrappers.scala:40)
[error]     at scala.collection.IterableLike.head(IterableLike.scala:106)
[error]     at scala.collection.IterableLike.head$(IterableLike.scala:105)
[error]     at scala.collection.AbstractIterable.head(Iterable.scala:54)
[error]     at scala.collection.TraversableLike.last(TraversableLike.scala:431)
[error]     at scala.collection.TraversableLike.last$(TraversableLike.scala:430)
[error]     at scala.collection.AbstractTraversable.last(Traversable.scala:104)
[error]     at play.api.i18n.DefaultMessagesApi.$anonfun$apply$3(Messages.scala:473)
[error]     at scala.Option.getOrElse(Option.scala:121)
[error]     at play.api.i18n.DefaultMessagesApi.apply(Messages.scala:473)
[error]     at play.i18n.MessagesApi.get(MessagesApi.java:106)
[error]     at play.data.Form.lambda$errorsAsJson$14(Form.java:750)
[error]     at java.util.ArrayList.forEach(ArrayList.java:1380)
[error]     at play.data.Form.errorsAsJson(Form.java:744)
[error]     at play.data.Form.errorsAsJson(Form.java:734)
[error]     at controllers.common.MyController.bindForm(MyController:20)
```

### PR DESCRIPTION
**Pull request by mkurz**
Fixes #8318 for the 2.7.x and master branch.

See https://github.com/playframework/playframework/issues/8318#issuecomment-455814157 for explanation.

### CODE REVIEW

**Comment by mkurz** [comment]
I also tested the reproducer project provided by @eximius313 in https://github.com/playframework/playframework/issues/8318#issuecomment-381346036 with this patch and can confirm that the problem is fixed there as well.

File: framework/src/play-java-forms/src/test/java/play/mvc/HttpFormsTest.java [hunk 1]
```diff
@@ -64,6 +67,12 @@ private JavaContextComponents contextComponents(Application app) {
             (Class[])null, app.injector().instanceOf(MessagesApi.class), app.injector().instanceOf(Formatters.class), app.injector().instanceOf(ValidatorFactory.class), app.injector().instanceOf(Config.class), formToCopy.lang().orElse(null));
     }
 
+    @After
+    public void after() {
+        // make sure we clean the current http context after each test run
+        Context.current.remove();
+    }
```
Comment (mkurz) [inline framework/src/play-java-forms/src/test/java/play/mvc/HttpFormsTest.java]: This `@After` should have been there before to make sure each test has a clean thread local. Right now each test sets the context, but no tests cleans it afterwards... I added it for safety.

**Comment by mkurz** [comment]
@marcospereira @dwijnand @renatocaval This is quite a tiny fix and I think it should be pretty safe to merge it into 2.7.x before you cut the final release. It would probably make it into 2.7.1 otherwise anyway so it shouldn't matter if we merge it now or then...

**Comment by dwijnand** [comment]
No worries.  I think we're going to take a week break and then finish the 2.7 release.

**Comment by mkurz** [comment]
Any reasons to *not* merge this?

**Comment by dwijnand** [comment]
Forward-ported in 9ba5a9d67a
Backported in #8972

**Comment by renatocaval** [comment]
> Can't remember why I felt the need to block it.

I think we were blocking it because we wanted to think a little bit if we wanted to have it in 2.7.0 or not. And we didn't want Mergify to kick in.

**Pull request merged**